applications/serial_lte_modem/src/slm_cmux.c

```diff
@@ -71,33 +80,84 @@ static void cmux_event_handler(struct modem_cmux *, enum modem_cmux_event event,
 	LOG_INF("CMUX %sconnected.", (event == MODEM_CMUX_EVENT_CONNECTED) ? "" : "dis");
 }
 
-static void init_dlci(size_t address, uint16_t recv_buf_size,
+static void init_dlci(size_t dlci_idx, uint16_t recv_buf_size,
 		      uint8_t recv_buf[static recv_buf_size])
 {
-	assert(ARRAY_SIZE(cmux.dlcis) >= address);
+	assert(ARRAY_SIZE(cmux.dlcis) > dlci_idx);
 
-	struct cmux_dlci *const dlci = &cmux.dlcis[address - 1];
+	struct cmux_dlci *const dlci = &cmux.dlcis[dlci_idx];
 	const struct modem_cmux_dlci_config dlci_config = {
-		.dlci_address = address,
+		.dlci_address = dlci_idx + 1,
 		.receive_buf = recv_buf,
 		.receive_buf_size = recv_buf_size
 	};
 
 	dlci->pipe = modem_cmux_dlci_init(&cmux.instance, &dlci->instance, &dlci_config);
-	dlci->recv_buf = recv_buf;
-	dlci->recv_buf_size = recv_buf_size;
-	dlci->address = address;
+	dlci->address = dlci_config.dlci_address;
 
 	modem_pipe_attach(dlci->pipe, dlci_pipe_event_handler, dlci);
 }
 
+static int cmux_write_at_channel(const uint8_t *data, size_t len)
+{
+	struct modem_pipe *const at_dlci_pipe = cmux.dlcis[AT_CHANNEL].pipe;
+	int ret = modem_pipe_transmit(at_dlci_pipe, data, len);
+	const int sent_len = MAX(0, ret);
+
+	if (ret == len) {
+		return 0;
+	}
+	if (ret >= 0) {
+		ret = 1;
+	}
+	LOG_ERR("Sent %u out of %u AT data bytes. (%d)", sent_len, len, ret);
+	return ret;
+}
+
+static void close_pipe(struct modem_pipe **pipe)
+{
+	if (*pipe) {
+		modem_pipe_close_async(*pipe);
+		modem_pipe_release(*pipe);
```

Comment:
As the above call is asynchronous, what happens if there is more data coming after this?

Reply:
This (`close_pipe()`) is only called from `cmux_stop()` when the setup of CMUX has failed (because otherwise only the UART pipe is closed), so there cannot be any data running through the pipes.

Though, in the hypothetical case you are mentioning, the `modem_pipe` module will see that the pipe has been released (`callback` has been set to `NULL`) and simply won't pass events forward (e.g. `MODEM_PIPE_EVENT_RECEIVE_READY`).